File: promotion/coupon-service/src/main/java/org/study/couponservice/repository/CouponRepository.java
package org.study.couponservice.repository;

import jakarta.persistence.LockModeType;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Lock;
import org.springframework.data.jpa.repository.Query;
import org.study.couponservice.domain.Coupon;

import java.util.Optional;

public interface CouponRepository extends JpaRepository<Coupon, Long> {

    Optional<Coupon> findByIdAndUserId(Long id, Long userId);

    @Query("SELECT COUNT(c) FROM Coupon c WHERE c.couponPolicy.id = :policyId")
    Long countByCouponPolicyId(Long policyId);

    Page<Coupon> findByUserIdAndStatusOrderByCreatedAtDesc(Long userId, Coupon.Status status, Pageable pageable);

    @Lock(LockModeType.PESSIMISTIC_WRITE)
    @Query("SELECT c FROM Coupon c WHERE c.id = :id")
    Optional<Coupon> findByIdWithLock(Long id);
}

File: promotion-kotlin/coupon-service-kotlin/src/main/kotlin/com/study/couponservice/repository/CouponRepository.kt
package com.study.couponservice.repository

import com.study.couponservice.domain.Coupon
import io.lettuce.core.dynamic.annotation.Param
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Query
import java.util.*

interface CouponRepository: JpaRepository<Coupon, Long> {
    fun findByIdAndUserId(id: Long, userId: Long): Optional<Coupon>

    @Query("SELECT COUNT(c) FROM Coupon c WHERE c.couponPolicy.id = :policyId")
    fun countByCouponPolicyId(@Param("policyId") policyId: Long): Long
}
